### DISCUSSION TITLE
Function dunders definition

### DISCUSSION THREAD

**Discussion by hmc-cs-mdrissi** [microsoft/pyright#3032] [Q&A] [answered]
Where do function dunder type definitions come from? Examples being,

```python
def foo(x: int) -> bool:
  ...

foo.__annotations__
foo.__signature__
foo.__code__
...
```

I'm unsure if there's a pyright bug or a typeshed bug as `foo.__signature__` is an error, but that dunder exists. Is it this [stub](https://github.com/python/typeshed/blob/master/stdlib/builtins.pyi#L792)?  The comment,

`# Doesn't exist at runtime, but deleting this breaks mypy.`

makes me unsure if that's a mypy specific stub or if other type checkers use it.

Broader question are there any special builtin types that don't have stubs in typeshed? Or should I expect all of them to be in builtins.pyi?

**Comment by erictraut** [comment, marked as answer]
> Where do function dunder type definitions come from?

There's unfortunately not a single answer to this question. It depends on which dunders you're talking about. Pyright tends to use definitions from typeshed wherever possible. For example, for functions (as in your example above), it uses the dunders from the `function` class in `builtins.pyi`. For modules, classes and functions, pyright's binder implicitly adds certain dunder symbols that are documented to exist but are not declared in typeshed stubs. You can find these by searching for `_addImplicitSymbolToCurrentScope` in `binder.ts`. Here's a [link](https://github.com/microsoft/pyright/blob/6566cb89d547d3feca22201a9494fc5a4b9c00bf/packages/pyright-internal/src/analyzer/binder.ts#L249).
